### PR TITLE
moe/engine: guard forward() against num_tokens exceeding scratch capacity

### LINKED ISSUE #18: moe/engine: forward() can overflow router scratch when num_tokens > max_tokens_ (OOB device writes)
## Summary

`MoEEngineImpl` allocates its router/top-k scratch once, sized for `max_tokens_ = 4096`:

```cpp
cudaMalloc(&d_logits_,  (size_t)max_tokens_ * cfg.num_experts * sizeof(float));
cudaMalloc(&d_ids_,     (size_t)max_tokens_ * cfg.top_k * sizeof(int));
cudaMalloc(&d_weights_, (size_t)max_tokens_ * cfg.top_k * sizeof(float));
```

But `forward(input, output, num_tokens, layer, stream)` passes the caller's `num_tokens` straight into the kernels with no check against `max_tokens_`:

```cpp
launch_moe_router_gemm(input, w.router_w, d_logits_, num_tokens, H, E, stream);   // writes num_tokens*E into d_logits_
launch_moe_router(d_logits_, d_ids_, d_weights_, d_counts_, num_tokens, E, K, ...); // writes num_tokens*K into d_ids_/d_weights_
```

If a caller ever routes a batch with `num_tokens > max_tokens_`, these kernels write past the allocated scratch — out-of-bounds device writes (memory corruption / crash), with no diagnostic.

## Location

`moe/src/engine.cpp` — `MoEEngineImpl::forward()`

## Why it matters

The current single-token decode path (`num_tokens = 1`) never exceeds the capacity, so this is latent today — but nothing enforces the invariant, and a future prefill/batched caller would silently corrupt memory. The engine should fail safely against its own documented scratch sizing.

## Proposed fix

Guard `forward()` before launching any kernel:

```cpp
if (num_tokens <= 0) return;
if (num_tokens > max_tokens_) {
    fprintf(stderr, "[moe] forward: num_tokens %d exceeds scratch capacity %d — skipping\n", num_tokens, max_tokens_);
    return;
}
```

## Impact

Robustness / memory-safety: turns a silent OOB device write into a clear, diagnosed refusal. Valid batches (num_tokens <= max_tokens_) are unaffected.

### PR DESCRIPTION
**Pull request by joaovictor712**
## Summary

Guard `MoEEngineImpl::forward()` so a batch larger than the engine's router scratch can't overflow it.

Fixes #18

**Why.** The router/top-k scratch (`d_logits_`, `d_ids_`, `d_weights_`) is allocated once for `max_tokens_ = 4096`, but `forward()` passed the caller's `num_tokens` into the kernels with no bound check. A batch with `num_tokens > max_tokens_` would write past those allocations — out-of-bounds device writes (memory corruption / crash). The current single-token decode path never hits it, but nothing enforced the invariant.

**Change.** Before launching any kernel, no-op on `num_tokens <= 0` and refuse (with a diagnostic) on `num_tokens > max_tokens_`:

```cpp
if (num_tokens <= 0) return;
if (num_tokens > max_tokens_) {
    fprintf(stderr, "[moe] forward: num_tokens %d exceeds scratch capacity %d — skipping\n", num_tokens, max_tokens_);
    return;
}
```

## Proof of speedup

- [ ] Tested on **RTX 5090** (`sm_120`)

N/A — this is a **robustness/memory-safety-only** change to the MoE host engine, not a perf change. Valid batches (`num_tokens <= max_tokens_`, e.g. the single-token decode path) take exactly the same code path as before — the guard only triggers on out-of-range inputs that previously caused OOB device writes. No decode-throughput impact.

**Benchmark log** (before → after):

```text
N/A — host-side bounds guard, no kernel/perf change. Valid-batch behavior unchanged.
```

| | decode tok/s |
|---|--:|
| before | N/A |
| after  | N/A |

## Scope

Single focused change in `moe/src/engine.cpp`; one fix per PR per CONTRIBUTING.md, scoped to `moe/`.

### CODE REVIEW

**Comment by ai-hpc** [comment]
## ⚪ sparkinfer auto-eval — `2bc0b87`

| metric | value |
|---|---|
| **label** | `eval:none` |
| decode | 186.29 tok/s |
| vs frontier | 187.61 tok/s → +0.0% (-1.3) |
| correctness | top-1 98.0% · KL 0.1434 |

Within the significance gate — no *verified* speedup over the current frontier.

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._
